### ISSUE TITLE
Variable columnNames is not available in Jython

### ISSUE THREAD

**Issue by wetneb** [OpenRefine/OpenRefine#4481]
The variable columnNames is not available in Jython, whereas it is available in GREL. It is useful to iterate over all cells in a row, for instance with this expression:
`forEach(row.columnNames,cn,cells[cn].value)`
(see https://groups.google.com/g/openrefine/c/gV0bLUz7p7o)

### Proposed solution

We could make this variable available in Jython too.
It would also be good to [document](https://docs.openrefine.org/manual/jythonclojure) which variable names (and fields) are available in Jython, if this is really meant to differ from what is available in GREL.

### Alternatives considered

Make sure all variables and fields offered in GREL are also available in Jython and Clojure.

**Comment by Keetna** [comment]
Hi @wetneb  , Im new to opensource . Could you assign this task to me and assist me to get cleared on my first issue? Thanks.

**Comment by wetneb** [comment]
Hi @Keetna, thanks for volunteering!
We have a guide for new contributors here, let me know if you get stuck at any stage:
https://docs.openrefine.org/technical-reference/contributing#your-first-code-pull-request

**Comment by Binita-tech** [comment]
@wetneb, I would like to work on this issue, kindly assign it to me.

**Comment by Binita-tech** [comment]
I found columnNames variable in GREL here:
![image](https://user-images.githubusercontent.com/78761614/161423472-18da3c72-e001-458d-9da2-03cb996914a8.png)
Now, I have to make available the same variable in Jython
![image](https://user-images.githubusercontent.com/78761614/161423689-2e11f821-0f68-40db-83c2-f2a4d5b5342b.png)
Is this only documentaion work? Please correct me if I am wrong @wetneb

**Comment by antoine2711** [comment]
> Now, I have to make available the same variable in Jython

Yes @Binita-tech.

> Is this only documentaion work? Please correct me if I am wrong @wetneb 

It's both. Adding code so Jython can handle the variable, and adding the corresponding documentation.

For the documentation, you could also document all the variables available in Jython. But it is not required, another issue could be opened for that.

Regards,
   Antoine

**Comment by jenny-Musah** [comment]
Hi @wetneb, I am a new bee to open source and so I would just say I think would work
I think creating a temporary CSV file that contains all the variables in GREL and then we export the CSV file from the GREL so that the Jython or Clojure code retrieves the variable value when the read the file.

Do you think it would work

**Comment by wetneb** [comment]
It sounds rather complicated and I do not think this is likely to work. I think this issue should be rather simple to fix once you have reviewed the existing Jython integration: it is just about adding another parameter to the Python function that is generated before evaluation.

**Comment by jenny-Musah** [comment]
Oh okay....except that I don't know  where to find the code, I am still trying to understand the code base, please can I get the git hub file part to  this code, so that i can try to make a contribution.

**Comment by ostephens** [comment]
I'm not very familiar with the Jython integration but I think the code that @wetneb is referring to is here https://github.com/OpenRefine/OpenRefine/blob/ab59b30a4527e844dd7ab37f2e56845b12effee0/extensions/jython/src/com/google/refine/jython/JythonEvaluable.java#L99

**Comment by jenny-Musah** [comment]
@ostephens Thanks

**Comment by jenny-Musah** [comment]
@wetneb, if I understand what is going on, I should add a parameter called  coloumnName, to the function "JythonEvaluable",
The parameter should help the function loop through each cells, right?

**Comment by wetneb** [comment]
Adding a `columnNames` parameter in the JythonEvaluable sounds appropriate indeed. Good luck!

**Comment by MahmoudAbdelkhalek5o5o** [comment]
Hello @wetneb,
I would like to work on this issue, kindly assign it to me.

**Comment by tfmorris** [comment]
My testing indicates

grel:columnNames => null
grel:row.columnNames => array of column names
jython:return columnNames => NameError: global name 'columnNames' is not defined
jython:return row.columnNames => array of columnNames
jython:return row["columnNames"] => array of columnNames

so I think we have parity between GREL and Jython in this area. There is no `columnNames` variable at the top level for either language.

The current list of top level variables and their fields as available here: https://github.com/OpenRefine/OpenRefine/issues/2523#issuecomment-713835727

**Comment by thadguidry** [comment]
I just wanted to point out this little fact that - for the variable, we have our docs stating otherwise, with the non-plural form `columnName` and not plural `columnNames` , but then `row` gained a member field that IS plural with `columnNames` ?

![image](https://github.com/user-attachments/assets/8b190f14-7c68-474e-a1c0-1539205373d2)

**Comment by wetneb** [comment]
Yes, the difference in plural/singular seems to match the expected semantics:
* `columnName` means the name of the column the operation is run on - a *single* column
* `columnNames` means the name of all columns in the project - generally *multiple* columns

So I think that's consistent.

**Comment by tfmorris** [comment]
I've repeated the [analysis](https://github.com/OpenRefine/OpenRefine/issues/4481#issuecomment-2364762766) that I did a year ago and it still holds true, so I'm going to close this as invalid. Please feel free to reopen if you disagree.
